bin/p2-setup-runtime/main.go

```diff
@@ -0,0 +1,191 @@
+	pod := podFactory.NewLegacyPod(manifest.ID())
+	return pod, nil
+}
+
+func main() {
+	app.Version(version.VERSION)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+
+	pods.Log.Logger.Formatter = &logrus.TextFormatter{
+		DisableTimestamp: false,
+		FullTimestamp:    true,
+		TimestampFormat:  "23:59:59.999",
+	}
+	logger := pods.Log
+
+	// read manifest
+	manifest, err := manifest.FromURI(*manifestURI)
+	if err != nil {
+		logger.WithError(err).WithField("manifest", *manifestURI).Fatalln("Unable to parse manifest file")
+	}
+
+	// list of hook names for which failure is fatal
+	hooksRequired := []string{}
+
+	// setup hooks context
+	hookLogger := logging.DefaultLogger
+	hookLogger.Logger.SetLevel(logrus.ErrorLevel)
+	auditLogger := hooks.NewFileAuditLogger(&hookLogger)
+	hookContext := hooks.NewContext(*hooksDir, *podRoot, &hookLogger, auditLogger)
+
+	logger = logger.WithField("app", manifest.ID())
+
+	// initialize pod
+	pod, err := initPod(manifest)
+	if err != nil {
+		logger.WithError(err).Fatalln("Failure initializing pod environment")
+
+	}
+
+	launchables, err := pod.Launchables(manifest)
+	if err != nil {
+		logger.WithError(err).Fatalln("Failure getting launchables from manifest")
+	}
+
+	if len(launchables) == 0 {
+		logger.Fatalln("There are no launchables in manifest")
+	}
+
+	//run hooks BeforeInstall
+	logger.Infoln("Running before_install hooks")
+	err = hookContext.RunHookType(hooks.BeforeInstall, pod, manifest, hooksRequired)
+	if err != nil {
+		logger.WithError(err).Fatalln("Failure running before_install hook(s)")
+	}
+
+	// run hooks AfterInstall
+	logger.Infoln("Running after_install hooks")
+	err = hookContext.RunHookType(hooks.AfterInstall, pod, manifest, hooksRequired)
+	if err != nil {
+		logger.WithError(err).Fatalln("Failure running after_install hook(s)")
+	}
+
+	// run hooks BeforeLaunch
+	logger.Infoln("Running before_launch hooks")
+	err = hookContext.RunHookType(hooks.BeforeLaunch, pod, manifest, hooksRequired)
+	if err != nil {
+		logger.WithError(err).WithField("app", manifest.ID()).Fatalln("Failure running before_launch hook(s)")
+	}
+
+	// run all parts of pods.Launch except for pod.StartLaunchables
+
+	// write current manifest
+	logger.Infoln("Writing current manifest")
+	oldManifestTemp, err := pod.WriteCurrentManifest(manifest)
+	defer os.RemoveAll(oldManifestTemp)
+	if err != nil {
+		logger.WithError(err).WithField("app", manifest.ID()).Fatalln("Unable to write current manifest")
+	}
+
+	// run post activate
+	logger.Infoln("Running post-activate")
+	err = pod.PostActivate(launchables)
+	if err != nil {
+		logger.WithError(err).Fatalln("Failure running PostActivate for launchables")
+	}
+
+	// build startup services
+	logger.Infoln("Running build runit services")
+	err = pod.BuildRunitServices(launchables, manifest)
+	if err != nil {
+		logger.WithError(err).Fatalln("Failure building run scripts")
+	}
+
+	// run AfterLaunch hooks
+	logger.Infoln("Running after_launch hooks")
```

Comment:
why we are running after_launch hooks before lauching(starting) a pod, this part is useless

Reply:
I didn't think about this closely on my initial review, but I now have a similar question to Zac. What is the goal/desire for running after hooks? If the application isn't running, are there any after hooks that we care about? If we don't run after hooks, what are the potential ramifications?

Reply:
running all hooks is a p2 promise, in here we do all of the pod.Launch  except for starting of pod services.
P2-preparer runs after_launch hooks irrespective of return status from `pod.StartLaunchables`
https://github.com/square/p2/blob/master/pkg/pods/pod.go#L202 ,
https://github.com/square/p2/blob/master/pkg/preparer/orchestrate.go#L582-L607 

so i felt it is good to run post_launch hooks even if you do not start services rather than completely skipping them. This closely mimics how p2-preparer would have set the pod.

Reply:
I added all hooks that are part of p2-preparer's 'launch' process, after_uninstall is not part of it so did not add it here. Yes prune images hooks does not make any impact for p2-one-shot project. All I wanted is to keep expectations of p2-setup-runtime pod to be same as p2-preparer's way of doing things. p2-preparer runs after_launch hooks even if startup of services fail, that means it is okay to run after_launch hooks even if we do not start services.

Reply:
I agree with @akun1012 that we shouldn't be doing the after_launch hooks since we are not launching. If a future after_launch hook were added that depends upon the application being up and running, then p2-one-shot would get stuck on this step, which would probably cause some madness in the imaging process (where we want to use this).

Reply:
okay, Updated pr.